vendor-extra/ContentPageBundle/src/Controller/ContentController.php

```diff
@@ -36,37 +39,88 @@ public function __construct(
         $this->dispatcher = $dispatcher;
     }
 
-    public function __invoke(Request $request, string $id) : Response
+    public function __invoke(Request $request, ?string $id = null) : Response
     {
-        return $this->client
-            ->requestAsync(
-                'GET',
-                "{$this->service}/items/{$id}/versions/latest",
-                [
-                    'headers' => ['Accept' => 'application/xml'],
-                    'http_errors' => true,
-                ]
-            )
-            ->then(
-                function (ResponseInterface $response) use ($request) : Response {
-                    $document = FluentDOM::load((string) $response->getBody());
-
-                    $context = [
-                        'lang' => $request->getLocale(),
-                        'dir' => text_direction($request->getLocale()),
-                    ];
-
-                    $event = new CreateContentPageEvent($document, $context);
-                    $this->dispatcher->dispatch($event::NAME, $event);
-
-                    return new Response(
-                        $this->twig->render(
-                            $this->template,
-                            $event->getContext() + ['title' => $event->getTitle(), 'content' => $event->getContent()]
-                        )
-                    );
-                }
-            )
-            ->wait();
+        if (!$id) {
+            return $this->client
+                ->requestAsync(
+                    'GET',
+                    "{$this->service}/items",
+                    [
+                        'headers' => ['Accept' => 'application/xml'],
+                        'http_errors' => true,
+                    ]
+                )
+                ->then(
+                    function (ResponseInterface $response) use ($request) : Response {
+                        $document = FluentDOM::load((string) $response->getBody());
+                        $document->registerNamespace('libero', 'http://libero.pub');
+
+                        $context = [
+                            'lang' => $request->getLocale(),
+                            'dir' => text_direction($request->getLocale()),
+                        ];
+
+                        $list = array_map(function (Element $element) use ($request, $context) {
+                            return $this->client
+                                ->requestAsync(
+                                'GET',
+                                "{$element->getAttribute('service')}/items/{$element->getAttribute('id')}/versions/latest"
+                                )
+                                ->then(
+                                    function (ResponseInterface $response) use ($request, $context) : Document {
+                                        return FluentDOM::load((string) $response->getBody());
+                                    }
```

Comment:
As each item in the list could return libero or JATS XML do we need to trigger an event for each item response?

Reply:
Something like https://github.com/libero/browser/blob/master/vendor-extra/JatsContentBundle/src/EventListener/BodyListener.php should pick this up, which will cycle through the items and convert them to teasers. The handling of that part will be view converts in the `JatsContentBundle` or `LiberoContentBundle`. However, they'll be the full item rather than just the relevant `<front>`, so might need to go through more than one...

Reply:
If I understand you correctly, we can get away with triggering a single event for the list and then have a listener for the whole list within `LiberoContentBundle` but the conversion of each item into views will be delegated to the content bundle which recognises the XML namespace of that item.